File: android/Girlsfleet/app/src/main/java/com/circus/girlsfleet/Vector.java
package com.circus.girlsfleet;

import android.graphics.Rect;

public class Vector {
    public float[] xyz=new float[3];

    public Vector(float x,float y){
        setX(x);
        setY(y);
        setZ(0.0f);
    }
    public Vector(float x,float y,float z){
        setX(x);
        setY(y);
        setZ(z);
    }

    public Vector(Rect r){
        setX(r.centerX());
        setY(r.centerY());
        setZ(0.0f);
    }
    public float getX(){
        return xyz[0];
    }
    public void setX(float f){
        xyz[0]=f;
    }
    public float getY(){
        return xyz[1];
    }
    public void setY(float f){
        xyz[1]=f;
    }
    public float getZ(){
        return xyz[2];
    }
    public void setZ(float f){
        xyz[2]=f;
    }

    public String toString(){
        return String.format("Vector{x:%f,y:%f,z:%f}",getX(),getY(),getZ());
    }
    public static Vector add(Vector v1,Vector v2){
        return new Vector(v1.getX()+v2.getX(),v1.getY()+v2.getY(),v1.getZ()+v2.getZ());
    }
    public static Vector sub(Vector v1,Vector v2){
        return new Vector(v1.getX()-v2.getX(),v1.getY()-v2.getY(),v1.getZ()-v2.getZ());
    }
    public static Vector div(Vector v,float s){
        return new Vector(v.getX()/s,v.getY()/s,v.getZ()/s);
    }
}

File: android/Circus/app/src/main/java/com/personal/circus/Vector.kt
package com.personal.circus

import android.graphics.Rect

class Vector :Cloneable{

    private var m_xyz=FloatArray(3)

    var xyz:FloatArray
    get(){
        return m_xyz
    }
    set(f:FloatArray){
        m_xyz=f.clone()
    }

    companion object {
        @JvmStatic
        val zero = Vector()

        fun add(r:Vector,v1:Vector,v2:Vector):Vector{
            for(i in 0..r.m_xyz.size-1){
                r.m_xyz[i]=v1.m_xyz[i]+v2.m_xyz[i]
            }
            return r
        }
        fun sub(r:Vector,v1:Vector,v2:Vector):Vector{
            for(i in 0..r.m_xyz.size-1){
                r.m_xyz[i]=v1.m_xyz[i]-v2.m_xyz[i]
            }
            return r
        }
        fun div(r:Vector,v1:Vector,s:Float):Vector{
            for(i in 0..r.m_xyz.size-1){
                r.m_xyz[i]=v1.m_xyz[i]/s
            }
            return r

        }
    }
    constructor(){
        xyz= floatArrayOf(0.0f,0.0f,0.0f)
    }
    constructor(f:FloatArray){
        xyz=f.clone()
    }
    constructor(r: Rect){
        xyz[0]=r.centerX().toFloat()
        xyz[1]=r.centerY().toFloat()
        xyz[2]=0.0f
    }

    fun getX():Float{
        return xyz[0]
    }
    fun setX(f:Float){
        xyz[0]=f
    }
    fun getY():Float{
        return xyz[1]
    }
    fun setY(f:Float){
        xyz[1]=f
    }

    fun getZ():Float{
        return xyz[2]
    }
    fun setZ(i:Float){
        xyz[2]=i
    }

    override public fun clone(): Any {
        return Vector(m_xyz)
    }
}